pkg/backfill/backfill.go

```diff
@@ -0,0 +1,171 @@
+package backfill
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/alecthomas/units"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/tsdb"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+// copied from https://github.com/prometheus/prometheus/blob/2f54aa060484a9a221eb227e1fb917ae66051c76/cmd/promtool/tsdb.go#L361-L398
+
+func printBlocks(blocks []tsdb.BlockReader, writeHeader, humanReadable bool, output io.Writer) {
+	tw := tabwriter.NewWriter(output, 13, 0, 2, ' ', 0)
+	defer tw.Flush()
+
+	if writeHeader {
+		fmt.Fprintln(tw, "BLOCK ULID\tMIN TIME\tMAX TIME\tDURATION\tNUM SAMPLES\tNUM CHUNKS\tNUM SERIES\tSIZE")
+	}
+
+	for _, b := range blocks {
+		meta := b.Meta()
+
+		fmt.Fprintf(tw,
+			"%v\t%v\t%v\t%v\t%v\t%v\t%v\t%v\n",
+			meta.ULID,
+			getFormatedTime(meta.MinTime, humanReadable),
+			getFormatedTime(meta.MaxTime, humanReadable),
+			time.Duration(meta.MaxTime-meta.MinTime)*time.Millisecond,
+			meta.Stats.NumSamples,
+			meta.Stats.NumChunks,
+			meta.Stats.NumSeries,
+			getFormatedBytes(b.Size(), humanReadable),
+		)
+	}
+}
+
+func getFormatedTime(timestamp int64, humanReadable bool) string {
+	if humanReadable {
+		return time.Unix(timestamp/1000, 0).UTC().String()
+	}
+	return strconv.FormatInt(timestamp, 10)
+}
+
+func getFormatedBytes(bytes int64, humanReadable bool) string {
+	if humanReadable {
+		return units.Base2Bytes(bytes).String()
+	}
+	return strconv.FormatInt(bytes, 10)
+}
+
+type Iterator interface {
+	Next() error
+	Sample() (ts int64, v float64)
+	Labels() (l labels.Labels)
+}
+
+type IteratorCreator func() Iterator
+
+// this is adapted from  https://github.com/prometheus/prometheus/blob/2f54aa060484a9a221eb227e1fb917ae66051c76/cmd/promtool/backfill.go#L68-L171
+func CreateBlocks(input IteratorCreator, mint, maxt int64, maxSamplesInAppender int, outputDir string, humanReadable bool, output io.Writer) (returnErr error) {
+	blockDuration := tsdb.DefaultBlockDuration
+	mint = blockDuration * (mint / blockDuration)
+
+	db, err := tsdb.OpenDBReadOnly(outputDir, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		var mErr tsdb_errors.MultiError
+		mErr.Add(returnErr)
+		mErr.Add(db.Close())
+		returnErr = mErr.Err()
```

Comment:
that's interesting, i've never seen a return value get assigned inside a deferred function, but it makes sense 👍

Reply:
I first came to contact with it, when it was suggested in a code review. To be honest it blew my mind. The backfill code is an adapted fork from https://github.com/prometheus/prometheus/blob/master/cmd/promtool/backfill.go